### DISCUSSION TITLE
issues with the wholeBodyDynamics and yarplogger modules

### DISCUSSION THREAD

**Discussion by itsjjaavii** [robotology/community#689] [Q&A]
hi, I currently having trouble with some modules of the yarpmanager. For a start, the only way I can run the yarplogger is through two ways: if I write in a terminal `yarplogger --start --no_stop` or if I run the icubsrv node through the terminal, writing `yarp run --server /icubsrv`. I cannot execute the yarplogger directly in the yarpmanager. On the other hand, I also cannot run the wholeBodyDynamics and gravityCompensator modules. When I try to, the yarplogger return this:
![yarplogger-wBD](https://github.com/robotology/community/assets/135467740/d17accf3-bff3-4b89-a17c-3f9f0869c202)

And if I try to run the wholeBodyDynamics module through the terminal, it return this:
![terminal-wBD](https://github.com/robotology/community/assets/135467740/de20239e-5cdc-4b23-ba35-268b0f84a920)

Since I can't run the wholeBodyDynamics module, the gravityCompensator module won't run either. 
I already saw the discussion #661 , however, the robot does have the inertial device enabled (the robot that we have here is iCubValparaiso01). 

I appreciate any help :)

**Comment by Nicogene** [comment]
Hi @itsjjaavii,

> I cannot execute the yarplogger directly in the yarpmanager. 

And other GUIs such as `yarpview` runs fine via yarpmanager?

> On the other hand, I also cannot run the wholeBodyDynamics and gravityCompensator modules.

From the log seems that the `multipleanalogsensorsserver` that publishes the FT data of the robot is not running. Is the `yarprobotinterface` running? Can you post the `yarp name list` output?

**Reply by itsjjaavii** [reply]
Hi @Nicogene,
I just tried to execute the `yarpview `module via yarpmanager, that didn't work either. 
The `yarprobotinterface` is working correctly. 
The output of the `yarp name list` command is quite extensive, so I'll divided in two images: 
![yarp-name-list](https://github.com/robotology/community/assets/135467740/5b846af0-81e2-489c-a3e9-9c9de50fd854)
![yarp-name-list2](https://github.com/robotology/community/assets/135467740/29c0ca65-bcc7-40be-8db8-d8df6f80e992)

**Reply by Nicogene** [reply]
Hi @itsjjaavii 

> I just tried to execute the yarpview module via yarpmanager, that didn't work either.

This seems a problem of exporting the display over ssh, in your [`cluster-config.xml`](https://github.com/robotology/icub-main/blob/master/app/yarpmanager/conf/cluster-config.xml) you are using for deploying the yarpruns how is set `display` ?
It should have the value that is in the `DISPLAY` env variable, that you can inspect with

```
echo $DISPLAY
```

which is the name of your iCub? Because some old icub configuration files have not been updated since a while after the institution of [`robot-icebox`](https://github.com/robotology/robots-configuration/issues/613#issuecomment-1959220307)

**Reply by mrojasanchez** [reply]
Hi @Nicogene, after checking that environment variable `DISPLAY` was `:1.0`, we solved the issue with yarplogger not opening inside yarpmanager by setting `<node display=":1.0">` in cluster-config.xml.

![Screenshot from 2024-03-15 17-35-21](https://github.com/robotology/community/assets/49776685/b6128ef6-a740-47e7-ba22-546ecfd3746a)

However, we still have to solve the problem with the wholebodydynamics module.

Attached is a log file from yarplogger.
[yarprunlog_15_03_2024_17_44_59.log](https://github.com/robotology/community/files/14620619/yarprunlog_15_03_2024_17_44_59.log)

Sorry for the inconvenience.

**Reply by Nicogene** [reply]
Hi @mrojasanchez,
> after checking that environment variable DISPLAY was :1.0, we solved the issue with yarplogger not opening inside yarpmanager by setting <node display=":1.0"> in cluster-config.xml.

I am glad to hear that!

> However, we still have to solve the problem with the wholebodydynamics module.

This is because the ft sensors of the arms are not exposed via [`multipleanalogsensorsserver`](https://www.yarp.it/latest/classMultipleAnalogSensorsServer.html), if you tell me the name of your iCub I can check in `robot-configuration` if there are the correct xml files. You need to:

```bash
echo $YARP_ROBOT_NAME
```

In any case, you can get rid of the error disabling the read from the FT sensors using this parameter for `wholebodydynamics` : `--dummy_ft`

**Reply by mrojasanchez** [reply]
The output of ` echo $YARP_ROBOT_NAME ` throws `iCubValparaiso01`.

For context, we are setting up a backup laptop to avoid messing up the Dell laptop that came with the iCub, e.g. testing updates, or new applications that require to install upgraded version packages

Also, we previously had to repair the left thumb (associated with [this](https://github.com/robotology/icub-tech-support/issues/1664) post.) Could it be related in some way? 

Thanks for the help @Nicogene .

**Reply by Nicogene** [reply]
Looking at the error you are getting seems that the left arm FT is not exposed, but if you are using this version of `icub_all.xml` for running `yarprobotinterface` you should have it:
- https://github.com/robotology/robots-configuration/blob/master/iCubValparaiso01/icub_all.xml

**Reply by justdiandes** [reply]
Hi @Nicogene ! I'm Diego. I'm currently trying to set the back up laptop. The problems that we are facing right know appears to be related to the WholeBodyDynamics port that need to be initialized in the /icubsrv and the /icub-head. All the modules are working correctly in the main laptop so the problem does not appear to be in the icub-head but in the backup laptop itself, and also we took a look in the icub_all.xml configuration for the yarprobotinterface and it's all in order.  When we trried to turn on the wholeBody module the only ports available that are related to this are the following: 

![image](https://github.com/robotology/community/assets/121194422/e0522ad2-78ad-4292-b375-7c9aa1dc866e)


And then the module shuts down, because it can't connect the other yarp ports. 

Our hypothesis now is around the possibility that the module in the laptop is not compatible with the icub-head version so we wanted to know what version of icub-head is know working or if we need to update the version in the iCub. We currently have a robotology version from 2021. 
I also wanted to know if you have any recommendation in order to track this issue or if you need extra information on the outputs to take a look for a possible error, because I don't understand if there is any problem with the IP configuration in all the initialized ports that can cause the problem of not finding the correct configurations for all the connections needed to initialized the WholeBody.

Please let me know if you need any additional information and I really thank you in advance!

**Reply by Nicogene** [reply]
>  if there is any problem with the IP configuration in all the initialized ports that can cause the problem of not finding the correct configurations for all the connections needed to initialized the WholeBody.

I would exclude this because the other modules seems able to communicate between `icub-head` and `icubsrv`.

> Our hypothesis now is around the possibility that the module in the laptop is not compatible with the icub-head version so we wanted to know what version of icub-head is know working or if we need to update the version in the iCub. We currently have a robotology version from 2021.

Some time ago we implemented some breaking changes in some `icub-main` modules, but it may be that different versions between the head and the laptop can give problems, but since iKinGazeCtrl seems to run fine I don't think it is the case.

Have you tried to run `wholeBodyDynamics` with `--dummy_ft`?

If the error you are facing is the same you posted earlier, please do this check after running `yarprobotinterface`:

```
yarp name list | grep /icub/left_arm/FT/measures:o
```


If the output is empty the problem is that the FT is not working/exposed, at that point you can:
-  run `wholeBodyDynamics` with `--dummy_ft`
or
- try to fix the FT

**Reply by justdiandes** [reply]
I checked on the `yarp name list` output to make sure if there's something but it doesn't appear any info so I tried to execute the module with `--dumy_ft` and it worked.

Now I wanted to know if this info is useful in some way 'cause now I wanted to know how to track this issue in order to setup this laptop. For context in this problem, the main laptop used for the iCub is working correctly so we believe that if there's any problem with the FT sensor itself, this could lead to an error in both laptops, but it only appear to fail in the backup one, so there's probably some configuration problem rather than a FT sensor problem.

**Reply by Nicogene** [reply]
The `yarprobotinterface` and the FT sensors devices runs on `icub-head`, the fact that with `yarp name list` you are not seeing the FT sensors ports means that something is not working on `yarprobotinterface` side.

Could you post the log of the `yarprobotinterface` from the `yarplogger`?
